pkg/didcomm/messenger/messenger.go

```diff
@@ -7,46 +7,155 @@ SPDX-License-Identifier: Apache-2.0
 package messenger
 
 import (
+	"encoding/json"
 	"errors"
+	"fmt"
 
+	"github.com/google/uuid"
+
+	"github.com/hyperledger/aries-framework-go/pkg/common/log"
 	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
 	"github.com/hyperledger/aries-framework-go/pkg/didcomm/dispatcher"
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+const (
+	messengerStore = "messenger_store"
+
+	jsonID             = "@id"
+	jsonThread         = "~thread"
+	jsonThreadID       = "thid"
+	jsonParentThreadID = "pthid"
 )
 
+var logger = log.New("aries-framework/didcomm/messenger")
+
+// record is an internal structure and keeps payload about inbound message
+type record struct {
+	MyDID    string
+	TheirDID string
+	ThreadID string
+}
+
 // Provider contains dependencies for the Messenger
 type Provider interface {
 	OutboundDispatcher() dispatcher.Outbound
+	StorageProvider() storage.Provider
 }
 
 // Messenger describes the messenger structure
 type Messenger struct {
+	store      storage.Store
 	dispatcher dispatcher.Outbound
 }
 
 // NewMessenger returns a new instance of the Messenger
-func NewMessenger(ctx Provider) *Messenger {
+func NewMessenger(ctx Provider) (*Messenger, error) {
+	store, err := ctx.StorageProvider().OpenStore(messengerStore)
+	if err != nil {
+		return nil, fmt.Errorf("open store: %w", err)
+	}
+
 	return &Messenger{
+		store:      store,
 		dispatcher: ctx.OutboundDispatcher(),
-	}
+	}, nil
 }
 
 // HandleInbound handles all inbound messages
-// function behavior is flexible and can be modified by providing options
-func (*Messenger) HandleInbound(msg service.DIDCommMsgMap, myDID, theirDID string) error {
-	return nil
+func (m *Messenger) HandleInbound(msg service.DIDCommMsgMap, myDID, theirDID string) error {
+	// an incoming message cannot be without id
+	if msg.ID() == "" {
+		return errors.New("message-id is absent and can't be processed")
+	}
+
+	// get message threadID
+	thID, err := msg.ThreadID()
+	if err != nil {
+		// since we are checking ID above this should never happen
+		// even if ~thread decorator is absent the message ID should be returned as a threadID
+		return fmt.Errorf("threadID: %w", err)
+	}
+
+	// saves message payload
+	return m.saveRecord(msg.ID(), record{MyDID: myDID, TheirDID: theirDID, ThreadID: thID})
 }
 
 // Send sends the message by starting a new thread.
 func (m *Messenger) Send(msg service.DIDCommMsgMap, myDID, theirDID string) error {
+	fixIfMalformed(msg)
+
+	if msg[jsonThread] != nil {
+		logger.Warnf("do not pass message %s, it will be ignored with the next change", jsonThread)
```

Comment:
message?

Reply:
@troyronda --> he meant to warn `do not pass message ~thread, it will be ignored with the next change`

@soluchok : It would be nicer if you can provide detailed comment on this function, something about using decorators, dos and don'ts would be nice. Just to avoid incorrect use of this function while replying.

And, are you checking data integrity here? something like, what if someone tries to send same message with same thread ID, sender order, ID collision etc? if not, then please mention it in comments.

Reply:
Changed to `do not pass message with ~thread decorator`